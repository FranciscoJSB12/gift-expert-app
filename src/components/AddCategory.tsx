import { ChangeEvent, FormEvent, useState } from 'react';

interface Props {
  onAddCategory: (value: string) => void;
}

export const AddCategory = ({ onAddCategory }: Props) => {
  const [inputValue, setInputValue] = useState('');

  const onChangeInput = (e: ChangeEvent<HTMLInputElement>): void => {
    setInputValue(e.target.value);
  };

  const onSubmitForm = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const trimmedValue = inputValue.trim();

    if (trimmedValue.length <= 1) return;

    onAddCategory(trimmedValue);

    setInputValue('');
  };

  return (
    <form
      onSubmit={onSubmitForm}
      aria-label='form'
    >
      <input
        type='text'
        placeholder='Buscar gifs'
        value={inputValue}
        onChange={onChangeInput}
      />
    </form>
  );
};
