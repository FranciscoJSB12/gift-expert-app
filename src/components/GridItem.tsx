interface Props {
  title: string;
  url: string;
}

export const GridItem = ({ title, url }: Props) => {
  return (
    <div className='card'>
      <img
        src={url}
        alt={title}
      />
      <p>{title}</p>
    </div>
  );
};
