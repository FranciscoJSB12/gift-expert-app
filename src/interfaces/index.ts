export * from './giphy-api-response.interface';
export * from './gif.interface';