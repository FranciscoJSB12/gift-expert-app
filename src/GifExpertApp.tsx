import { useState } from 'react';
import { AddCategory, GifGrid } from './components';

export const GifExpertApp = () => {
  const [categories, setCategories] = useState<string[]>(['One Punch']);

  const onAddCategory = (value: string): void => {
    if (categories.includes(value)) return;

    setCategories([...categories, value]);
  };

  return (
    <>
      <h1>GifExpertApp</h1>
      <AddCategory onAddCategory={onAddCategory} />
      {categories.map(category => (
        <GifGrid
          key={category}
          category={category}
        />
      ))}
    </>
  );
};
