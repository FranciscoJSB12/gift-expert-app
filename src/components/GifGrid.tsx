import { useFetchGifs } from '../hooks/useFetchGifs';
import { GridItem } from './GridItem';

interface Props {
  category: string;
}

export const GifGrid = ({ category }: Props) => {
  const { images, isLoading, error } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>

      {!isLoading && error && (
        <p>No se pudo cargar la información, prueba de nuevo...</p>
      )}

      {isLoading && <p>Cargando...</p>}

      {images.length === 0 && <p>No se encontró resultados</p>}

      <div className='card-grid'>
        {images.map(image => (
          <GridItem
            key={image.id}
            {...image}
          />
        ))}
      </div>
    </>
  );
};
