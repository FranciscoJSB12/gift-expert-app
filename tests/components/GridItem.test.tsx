import { test, describe, expect, afterEach } from 'vitest';
import { cleanup, render, screen } from '@testing-library/react';
import { GridItem } from '../../src/components';

describe('Testing <GridItem />', () => {
  const title = 'One Punch Man GIF';
  const url =
    'https://media2.giphy.com/media/VXJWhaO7afRe/giphy.gif?cid=9f017724sa8fjxmpligfbyoguybcvci91u0m7p7x2i7gfa7k&ep=v1_gifs_search&rid=giphy.gif&ct=g';

  afterEach(cleanup);

  test('Should match the snapshot', () => {
    const { container } = render(
      <GridItem
        title={title}
        url={url}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('Should render the image with its URL and ALT', () => {
    render(
      <GridItem
        title={title}
        url={url}
      />
    );

    // expect(screen.getByRole<HTMLImageElement>('img').src).toBe(url);
    // expect(screen.getByRole<HTMLImageElement>('img').alt).toBe(title);

    const { alt, src } = screen.getByRole<HTMLImageElement>('img');

    expect(src).toBe(url);
    expect(alt).toBe(alt);
  });

  test('Should display the title', () => {
    render(
      <GridItem
        title={title}
        url={url}
      />
    );

    expect(screen.getByText(title)).toBeTruthy();
  });
});
