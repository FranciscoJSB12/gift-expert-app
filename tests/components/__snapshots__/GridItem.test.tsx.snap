// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing <GridItem /> > Should match the snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="One Punch Man GIF"
      src="https://media2.giphy.com/media/VXJWhaO7afRe/giphy.gif?cid=9f017724sa8fjxmpligfbyoguybcvci91u0m7p7x2i7gfa7k&ep=v1_gifs_search&rid=giphy.gif&ct=g"
    />
    <p>
      One Punch Man GIF
    </p>
  </div>
</div>
`;
